src/main/java/com/codeborne/selenide/commands/ancestor/AncestorWithAttributeAndValueRule.java

```diff
@@ -0,0 +1,25 @@
+package com.codeborne.selenide.commands.ancestor;
+
+import static java.lang.String.format;
+
+public class AncestorWithAttributeAndValueRule extends SelectorValidation implements AncestorRule {
+
+  private String xpath;
```

Comment:
**Problem:**
`AncestorWithAttributeAndValueRule` holds the found `xpath` as a field which can be concurrently written by parallel threads (instance of `AncestorWithAttributeAndValueRule` itself is stored as a static variable in `AncestorRuleEngine`). 

**Proposed solution:**
Instead of holding the found `xpath` in a field and returning a `boolean`, method `evaluate` could **return**  `Optional<AncestorResult>`.

Reply:
I'm not sure I understood. I change `evaluate()` method in rules like this:

```java
  public Optional<AncestorResult> evaluate(String selector, int index) {
    if (isAttribute(selector) && containsAttributeValue(selector)) {
      int endIndex = selector.indexOf("=");
      String attribute = selector.substring(1, endIndex);
      String attributeValue = selector.substring(endIndex + 1, selector.length() - 1);
      this.xpath = format("ancestor::*[@%s='%s'][%s]", attribute, attributeValue, index);
    }
    return Optional.of(new AncestorResult(xpath));
  }
```

Then there rule engine class:

```java
  public AncestorResult process(String selector, int index) {
    AncestorRule ancestorRule = rules
      .stream()
      .filter(rule -> rule.evaluate(selector, index).isPresent())
      .findFirst()
      .orElseThrow(() -> new IllegalArgumentException("Selector does not match any rule"));
    return ancestorRule.getAncestorResult();
  }
```

But the `xpath` is `null`.
Could you please help me?

Reply:
Yes, just remove field `xpath` and method `getAncestorResult()` from `AncestorRule`. Use only the value returned by method `evaluate`.

Reply:
Thanks. Updated.